firestore-leaderboard/extension/functions/src/index.ts

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as admin from "firebase-admin";
+import * as functions from "firebase-functions";
+
+import { FieldValue } from "firebase-admin/firestore";
+import { getEventarc } from "firebase-admin/eventarc";
+
+import config from "./config";
+import * as logs from "./logs";
+
+enum ChangeType {
+  CREATE,
+  DELETE,
+  UPDATE,
+}
+
+// Initialize the Firebase Admin SDK
+admin.initializeApp();
+const db = admin.firestore();
+
+logs.init(config);
+
+/** Setup EventArc Channels */
+const eventChannel =
+  process.env.EVENTARC_CHANNEL &&
+  getEventarc().channel(process.env.EVENTARC_CHANNEL, {
+    allowedEventTypes: process.env.EXT_SELECTED_EVENTS,
+  });
+
+const getChangeType = (
+  change: functions.Change<admin.firestore.DocumentSnapshot>
+): ChangeType => {
+  if (!change.after.exists) {
+    return ChangeType.DELETE;
+  }
+  if (!change.before.exists) {
+    return ChangeType.CREATE;
+  }
+  return ChangeType.UPDATE;
+};
+
+const publishEvent = async (leaderboardPath: string) => {
+  if (eventChannel) {
+    await eventChannel.publish({
+      type: `firebase.extensions.firestore-leaderboard.v1.updated`,
+      data: {
+        documentPaths: leaderboardPath,
+      },
+    });
+  }
+};
+
+const addEntryLeaderboardDocument = async (
+  change: functions.Change<admin.firestore.DocumentSnapshot>
+): Promise<void> => {
+  console.log(`Start addEntryLeaderboardDocument()`);
+
+  const leaderboardCollectionRef = db
+    .collection(config.leaderboardCollectionPath)
+    .doc(config.leaderboardName);
+
+  await db.runTransaction(async (transaction) => {
+    const user_id = change.after.ref.id;
+    const entryData = {
+      [config.scoreFieldName]: change.after.get(config.scoreFieldName),
+      [config.userNameFieldName]: change.after.get(config.userNameFieldName),
+    };
+    transaction.set(
+      leaderboardCollectionRef,
+      { [user_id]: entryData },
+      { merge: true }
+    );
+    publishEvent(leaderboardCollectionRef.path);
+    console.log(`End addEntryLeaderboardDocument()`);
+    return Promise.resolve();
+  });
+};
+
+const deleteEntryLeaderboardDocument = async (
```

Comment:
Hmm if you delete an entry, I would expect the leaderboard to update the ranks of the rest of the entries, and fill it in with the next highest user. I think simply removing the entry could lead to inaccurate results?

Let's say top 1-100 scores are > 10
Top scorer 101 has 9
We delete one of the top 100 scorers
Another user scores 8

The latest user will get into the leaderboard even though there is someone who has a higher score than them.

I wonder if we should remove the deleteEntry use case or add a caveat in the documentation explaining this, I personally was a bit confused with the end result.

Reply:
👍

Reply:
Delete entry from the leaderboard should not matter too much, since next time when a new entry is added, it would be refreshed.

Added the logic to make sure leaderboard will not get oversized.

Reply:
Is this scenario possible?

Let's assume Leaderboard is full (has 10 entries)

Let’s say there is a user, u1, that has score 10 (they are not in the leaderboard)

Now, the 10th entry gets deleted from leaderboard

A new user, u2, scores 9

u2 will get added to leaderboard even though u1 should be in the leaderboard